### PR TITLE
resource booking

### PR DESCRIPTION
**Pull request by georgehrke**
ToDo:
- [x] add proper testing instructions
- [x] unit test for Managers / TimedJob

### CODE REVIEW

File: apps/dav/lib/CalDAV/CalendarRoot.php [hunk 1]
```diff
@@ -27,4 +27,16 @@ class CalendarRoot extends \Sabre\CalDAV\CalendarRoot {
...
+
+	function getName() {
+		if ($this->principalPrefix === 'principals/calendar-resources' ||
+			$this->principalPrefix === 'principals/calendar-rooms') {
+			$parts = explode('/', $this->principalPrefix);
+
+			return $parts[1];
+		}
+
+		\OC::$server->getLogger()->error(parent::getName());
```
Comment (georgehrke) [inline apps/dav/lib/CalDAV/CalendarRoot.php]: todo: remove

File: lib/public/Calendar/Resource/IManager.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	/**
+	 * @return IBackend[]
+	 * @since 14.0.0
+	 */
+	public function getBackends():array;
+
+	/**
+	 * @param string $backendId
+	 * @return IBackend
```
Comment (georgehrke) [inline lib/public/Calendar/Resource/IManager.php]: missing since annotation

File: lib/public/Calendar/Resource/IBackend.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+namespace OCP\Calendar\Resource;
+use OCP\Calendar\BackendTemporarilyUnavailableException;
+
+/**
+ * Interface IBackend
+ *
+ * @package OCP\Calendar\Resource
+ * @since 14.0.0
+ */
+interface IBackend {
```
Comment (georgehrke) [inline lib/public/Calendar/Resource/IBackend.php]: plenty of missing since annotation

File: lib/public/Calendar/Room/IBackend.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+namespace OCP\Calendar\Room;
+use OCP\Calendar\BackendTemporarilyUnavailableException;
+
+/**
+ * Interface IBackend
+ *
+ * @package OCP\Calendar\Room
+ * @since 14.0.0
+ */
+interface IBackend {
```
Comment (georgehrke) [inline lib/public/Calendar/Room/IBackend.php]: plenty of missing missing since annotation

File: lib/private/Calendar/Room/Manager.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	 * @return IBackend[]
+	 * @since 14.0.0
+	 */
+	public function getBackends():array {
+		return array_values($this->backends);
+	}
+
+	/**
+	 * @param string $backendId
+	 * @return IBackend|null
```
Comment (georgehrke) [inline lib/private/Calendar/Room/Manager.php]: missing since annotation

File: lib/private/Calendar/Resource/Manager.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	 * @return IBackend[]
+	 * @since 14.0.0
+	 */
+	public function getBackends():array {
+		return array_values($this->backends);
+	}
+
+	/**
+	 * @param string $backendId
+	 * @return IBackend|null
```
Comment (georgehrke) [inline lib/private/Calendar/Resource/Manager.php]: missing since annotation

**Comment by MorrisJobke** [comment]
> add proper testing instructions

🏓

**Comment by georgehrke** [comment]
To review this:
- install a resource booking backend like https://github.com/georgehrke/admin_resource_booking_database
(that backend is still in development, you can already use it to add / remove resources. You have to enter valid json into the group_restrictions field for now. I checked in all the compiled javascript files, so no need to run any build command)
- add resources / rooms:
5e9dff26-1368-4324-bb4a-a3bd79e77fc8
(all fields are mandatory!)
- open thunderbird
- schedule an event with one of the rooms / resources as attendee
- schedule another event on the same day and check if the time slot you previously selected is blocked in the Invite attendees dialog
- move events around to see that Free/Busy data for resources / rooms update properly
- delete events to see that time slots are freed again.

**Comment by rullzer** [comment]
Gave it a quick try and seems to do its job. Awesome.
I'll have a closer look at the code later.

**Comment by MorrisJobke** [comment]
CI: https://drone.nextcloud.com/nextcloud/server/8471

**Comment by juliushaertl** [comment]
I'm getting the following error when I try to add a resource:

Too few arguments to function OCA\\DAV\\BackgroundJob\\UpdateCalendarResourcesRoomsBackgroundJob::run(), 0 passed in \/var\/www\/html\/apps\/dav\/appinfo\/app.php on line 54 and exactly 1 expected

```
{"reqId":"wKQZfztGtnE0JcDGm9Ok","level":3,"time":"2018-06-27T17:02:07+00:00","remoteAddr":"192.170.0.1","user":"admin","app":"index","method":"POST","url":"\/index.php\/apps\/admin_resource_booking_database\/api\/resources","message":{"Exception":"ArgumentCountError","Message":"Too few arguments to function OCA\\DAV\\BackgroundJob\\UpdateCalendarResourcesRoomsBackgroundJob::run(), 0 passed in \/var\/www\/html\/apps\/dav\/appinfo\/app.php on line 54 and exactly 1 expected","Code":0,"Trace":[{"file":"\/var\/www\/html\/apps\/dav\/appinfo\/app.php","line":54,"function":"run","class":"OCA\\DAV\\BackgroundJob\\UpdateCalendarResourcesRoomsBackgroundJob","type":"->","args":[]},{"file":"\/var\/www\/html\/3rdparty\/symfony\/event-dispatcher\/EventDispatcher.php","line":212,"function":"{closure}","class":"OC_App","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"\/var\/www\/html\/3rdparty\/symfony\/event-dispatcher\/EventDispatcher.php","line":44,"function":"doDispatch","class":"Symfony\\Component\\EventDispatcher\\EventDispatcher","type":"->","args":[[{"__class__":"Closure"}],"*** sensitive parameter replaced ***","*** sensitive parameter replaced ***"]},{"file":"\/var\/www\/html\/apps-extra\/admin_resource_booking_database\/lib\/Controller\/ResourceBookingController.php","line":149,"function":"dispatch","class":"Symfony\\Component\\EventDispatcher\\EventDispatcher","type":"->","args":["*** sensitive parameter replaced ***"]},{"file":"\/var\/www\/html\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":166,"function":"createResource","class":"OCA\\AdminResourceBookingDatabase\\Controller\\ResourceBookingController","type":"->","args":[]},{"file":"\/var\/www\/html\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":99,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\AdminResourceBookingDatabase\\Controller\\ResourceBookingController"},"createResource"]},{"file":"\/var\/www\/html\/lib\/private\/AppFramework\/App.php","line":118,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\AdminResourceBookingDatabase\\Controller\\ResourceBookingController"},"createResource"]},{"file":"\/var\/www\/html\/lib\/private\/AppFramework\/Routing\/RouteActionHandler.php","line":47,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\AdminResourceBookingDatabase\\Controller\\ResourceBookingController","createResource",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"admin_resource_booking_database.resource_booking.createResource"}]},{"function":"__invoke","class":"OC\\AppFramework\\Routing\\RouteActionHandler","type":"->","args":[{"_route":"admin_resource_booking_database.resource_booking.createResource"}]},{"file":"\/var\/www\/html\/lib\/private\/Route\/Router.php","line":297,"function":"call_user_func","args":[{"__class__":"OC\\AppFramework\\Routing\\RouteActionHandler"},{"_route":"admin_resource_booking_database.resource_booking.createResource"}]},{"file":"\/var\/www\/html\/lib\/base.php","line":992,"function":"match","class":"OC\\Route\\Router","type":"->","args":["\/apps\/admin_resource_booking_database\/api\/resources"]},{"file":"\/var\/www\/html\/index.php","line":42,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"\/var\/www\/html\/apps\/dav\/lib\/BackgroundJob\/UpdateCalendarResourcesRoomsBackgroundJob.php","Line":87,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (X11; Linux x86_64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/67.0.3396.87 Safari\/537.36","version":"14.0.0.5"}
```

**Comment by georgehrke** [comment]
@juliushaertl Not sure why i didnt receive that error message, but it should be fixed now

**Review by MorrisJobke — APPROVED** [review]
Code looks good and I gave it a rough smoke test 👍

**Review by rullzer — APPROVED** [review]
Also my quick smoke test worked.

**Comment by almet** [comment]
Hey, I believe this has been merged, but I cannot find how to use this from my nextcloud server. Is there any documentation anywhere? Thanks!

**Comment by greenvespa** [comment]
I'm looking for this feature too, I thought it was added in Nextcloud 14 as described in this official announcement https://nextcloud.com/blog/nextcloud-14-now-available-with-video-verification-signaltelegram-2fa-support-improved-collaboration-and-gdpr-compliance/ but I can't find it in my nextcloud server. Am I missing something? Thanks!

**Comment by almet** [comment]
@greenvespa haven't found myself, it doesn't seem to be working ATM. Let me know if you find something!

**Comment by Shen** [comment]
What happened to this great feature? Still in development?

**Comment by fwolfst** [comment]
> What happened to this great feature? Still in development?
Calender 2.0 just released. I think now it will slowly become time to revive this (if it was actually sleeping).

**Comment by georgehrke** [comment]
> I think now it will slowly become time to revive this (if it was actually sleeping).

Please see https://github.com/nextcloud/server/issues/17961#issuecomment-554401645

**Pull request merged**